File: 0048. Rotate Image/Solution0048.java
public class Solution0048 {
    public void rotate(int[][] matrix) {
        int topRow = 0;
        int bottomRow = matrix.length - 1;
        int leftCol = 0;
        int rightCol = matrix[0].length - 1;
        while (topRow <= bottomRow && leftCol <= rightCol) {
            for (int i = 0; i < rightCol - leftCol; i++) {
                int temp = matrix[topRow][leftCol + i];
                matrix[topRow][leftCol + i] = matrix[bottomRow - i][leftCol];
                matrix[bottomRow - i][leftCol] = matrix[bottomRow][rightCol - i];
                matrix[bottomRow][rightCol - i] = matrix[topRow + i][rightCol];
                matrix[topRow + i][rightCol] = temp;
            }
            topRow++;
            bottomRow--;
            leftCol++;
            rightCol--;
        }
    }    
}

File: 0048. Rotate Image/Solution0048.cpp
# include<vector>
# include<iostream>
using namespace std;

class Solution0048 {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int level = 0;
        while (level < n / 2) {
            for (int i = level; i < n - 1 - level; i++) {
                int tmp = matrix[level][i];
                matrix[level][i] = matrix[n-1-i][level]; // 1- , 7|
                matrix[n-1-i][level] = matrix[n-1-level][n-1-i]; // 7\, 9-
                matrix[n-1-level][n-1-i] = matrix[i][n-1-level]; //9-. 3|
                matrix[i][n-1-level] = tmp;
            }
            level++;
        }
        return;
    }
};